src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs

```diff
@@ -597,131 +808,228 @@ protected Type TryGetType(string name)
                 paramMatches.Add(name, val);
             }
             return paramMatches;
-
+        }
+        
+        
+        public List<string> findClrReferences(string code)
+        {
+            var statements = new List<string>();
+            foreach (var line in code.Split(new[] {'\n', ';'}))
+            {
+                if (line.Contains("clr.AddReference"))
+                {
+                    statements.Add(line.Trim());
+                }
+            }
+            return statements;
         }
 
         /// <summary>
-        /// Find all import statements and import into scope.  If the type is already in the scope, this will be skipped.  
-        /// The ImportedTypes dictionary is 
+        /// Find all import statements and import into scope.  If the type is already in the scope, this will be skipped.
+        /// The ImportedTypes dictionary is
         /// </summary>
         /// <param name="code">The code to discover the import statements.</param>
         public void UpdateImportedTypes(string code)
         {
-            // look all import statements
-            var imports = FindBasicImportStatements(code)
-                .Union(FindTypeSpecificImportStatements(code))
-                .Union(FindAllTypeImportStatements(code));
-
-            // try and load modules into python scope
-            foreach (var import in imports)
+            //detect all lib references prior to attempting to import anything
+            var refs = findClrReferences(code);
+            foreach (var statement in refs)
             {
-                if (scope.ContainsVariable(import.Key))
+                int previousTries = 0;
+                badStatements.TryGetValue(statement, out previousTries);
+                if (previousTries > 3)
                 {
                     continue;
                 }
+                
                 try
                 {
-                    scope.Engine.CreateScriptSourceFromString(import.Value, SourceCodeKind.SingleStatement)
-                         .Execute(this.scope);
-                    var type = Type.GetType(import.Key);
-                    this.ImportedTypes.Add(import.Key, type);
+                    string libName = MATCH_FIRST_QUOTED_NAME.Match(statement).Groups[1].Value;
+                    if (!clrModules.Contains(libName))
+                    {
+                        if (statement.Contains("AddReferenceToFileAndPath"))
+                        {
+                            engine.CreateScriptSourceFromString(statement, SourceCodeKind.SingleStatement).Execute(scope);
+                            //it's an assembly path, don't check the current appdomain
+                            clrModules.Add(libName);
+                            continue;
+                        }
+                        
+                        if(AppDomain.CurrentDomain.GetAssemblies().Any(x => x.GetName().Name == libName))
```

Comment:
hmm, maybe it is worth making a map of assemblies and checking if the name exists there instead of this loop - it appears lookup might happen a lot.